content/learn/book/the-game-loop/time-and-timers.md

```diff
@@ -0,0 +1,277 @@
+number of updates between each frame.
+This will lead to visible jittering, lagginess and tiny speedups.
+
+To account for this, we need to distinguish between the logical and rendered position (and rotation, and sometimes scale) of
+game objects.
+This is a common problem and pattern for multiplayer games, but using a fixed timestep means that even single player
+games need to handle it.
+
+The core strategy is simple enough, but the devil is in the details:
+
+1. Create your own custom `GamePosition` type, holding the translation/rotation/scale information that your project needs.
+2. Modify and read this `GamePosition` type for all of your game logic and physics.
+3. Interpolate between the previous [`GlobalTransform`] and the `GamePosition`, based on the elapsed [`Fixed`] time.
+   1. This should occur after the final fixed update for the frame has run, but before rendering occurs.
+   2. You ultimately want to set the [`GlobalTransform`], but doing this correctly in the presence of hierarchies is hard.
+   3. As a result, modifying the local [`Transform`] and relying on transform propagation can be the least bad solution.
+
+Bevy does not currently offer any built-in functionality for this form of interpolation,
+but open source [ecosystem crates] are available to use and learn from.
+
+[ecosystem crates]: https://bevy.org/assets/
+[`GlobalTransform`]: https://docs.rs/bevy/latest/bevy/prelude/struct.GlobalTransform.html
+[`Transform`]: https://docs.rs/bevy/latest/bevy/prelude/struct.Transform.html
+
+## Timers and cooldowns
+
+Delta time is great for physics and animations,
+but what if we want to implement other time-driven gameplay logic?
+We might want to automatically generate cookie clicks every 5 seconds
+or add a 3 second cooldown for our fireball ability.
+
+The simplest of these tools is the [`Timer`].
+Timers in Bevy hold two [`Duration`]s:
+one to store the duration of the timer, and one which stores how much time has been elapsed.
+
+Timers have no inherent logic: they do not update themselves,
+or have any inherent way to add something like an "on completion callbacks".
+They are also not components: you cannot simply add a [`Timer`] to an entity.
+
+Instead, timers are intended to be wrapped inside of simple components,
+which are updated and polled for completion by systems.
+
+```rust
+# use bevy::prelude::*;
+#
+#[derive(Resource)]
+struct Cookies(u64);
+
+#[derive(Component)]
+struct AutomaticCookieClick {
+  // This should be initialized as a repeating timer
+  // ensuring it automatically resets
+  timer: Timer,
+  resources_gained: u64,
+}
+
+fn automatically_click_cookies(
+  mut cookies: ResMut<Cookies>, 
+  time: Res<Time>, mut query: Query<&mut AutomaticCookieClick>
+){
+  let delta_time = time.delta();
+
+  for mut cookie_clicker in query.iter_mut(){
+    cookie_clicker.timer.tick(delta_time);
+    if cookie_clicker.timer.just_finished(){
+      cookies.0 += cookie_clicker.resources_gained;
+    }
+  }
+}
```

Comment:
`jondolf fmt`
```suggestion
#[derive(Component)]
struct AutomaticCookieClick {
    // This should be initialized as a repeating timer
    // ensuring it automatically resets
    timer: Timer,
    resources_gained: u64,
}

fn automatically_click_cookies(
    mut cookies: ResMut<Cookies>, 
    mut query: Query<&mut AutomaticCookieClick>,
    time: Res<Time>,
) {
    let delta_time = time.delta();

    for mut cookie_clicker in query.iter_mut() {
        cookie_clicker.timer.tick(delta_time);
        if cookie_clicker.timer.just_finished() {
            cookies.0 += cookie_clicker.resources_gained;
        }
    }
}
```
Side note, I bet we could get CI to automatically check formatting for code snippets